modules/navier_stokes/src/fv/kernels/INSFVAdvectionBase.C

```diff
@@ -0,0 +1,178 @@
+               static_cast<std::string>(velocity_interp_method));
+
+  if (_nsfv_tid == 0)
+  {
+    auto & vec_of_coeffs_map = _rc_a_coeffs[&_nsfv_app];
+    vec_of_coeffs_map.resize(libMesh::n_threads());
+  }
+}
+
+const ADReal &
+INSFVAdvectionBase::rcCoeff(const Elem & elem, const ADReal & mu) const
+{
+  auto it = _rc_a_coeffs.find(&_nsfv_app);
+  mooseAssert(it != _rc_a_coeffs.end(),
+              "No RC coeffs structure exists for the given MooseApp pointer");
+  mooseAssert(_nsfv_tid < it->second.size(),
+              "The RC coeffs structure size "
+                  << it->second.size() << " is greater than or equal to the provided thread ID "
+                  << _nsfv_tid);
+  auto & my_map = it->second[_nsfv_tid];
+
+  auto rc_map_it = my_map.find(&elem);
+
+  if (rc_map_it != my_map.end())
+    return rc_map_it->second;
+
+  // Returns a pair with the first being an iterator pointing to the key-value pair and the second a
+  // boolean denoting whether a new insertion took place
+  auto emplace_ret = my_map.emplace(&elem, coeffCalculator(elem, mu));
+
+  mooseAssert(emplace_ret.second, "We should have inserted a new key-value pair");
+
+  return emplace_ret.first->second;
+}
+
+ADReal
+INSFVAdvectionBase::coeffCalculator(const Elem & elem, const ADReal & mu) const
+{
+  ADReal coeff = 0;
+
+  ADRealVectorValue elem_velocity(_u_var->getElemValue(&elem));
+
+  if (_v_var)
+    elem_velocity(1) = _v_var->getElemValue(&elem);
+  if (_w_var)
+    elem_velocity(2) = _w_var->getElemValue(&elem);
+
+  auto action_functor = [&coeff, &elem_velocity, &mu, this](const Elem & /*functor_elem*/,
+                                                            const Elem * const neighbor,
+                                                            const FaceInfo * const fi,
+                                                            const Point & surface_vector,
+                                                            Real coord,
+                                                            const bool /*elem_has_info*/) {
+    mooseAssert(fi, "We need a non-null FaceInfo");
+    ADRealVectorValue neighbor_velocity(_u_var->getNeighborValue(neighbor, *fi, elem_velocity(0)));
+    if (_v_var)
+      neighbor_velocity(1) = _v_var->getNeighborValue(neighbor, *fi, elem_velocity(1));
+    if (_w_var)
+      neighbor_velocity(2) = _w_var->getNeighborValue(neighbor, *fi, elem_velocity(2));
+
+    ADRealVectorValue interp_v;
+    Moose::FV::interpolate(Moose::FV::InterpMethod::Average,
+                           interp_v,
+                           elem_velocity,
+                           neighbor_velocity,
+                           *fi,
+                           neighbor == fi->neighborPtr());
+
+    ADReal mass_flow = _rho * interp_v * surface_vector;
+
+    coeff += -mass_flow;
+
+    // Now add the viscous flux
+    coeff += mu * fi->faceArea() * coord / (fi->elemCentroid() - fi->neighborCentroid()).norm();
+  };
```

Comment:
this functor should be mindful of the corrections to the a coefficient in Rhie Chow interpolation on the boundaries (eq. 15.124 ... 134 for walls and 15.136 for inlet, 139 and 142 for outlet).

In general, what is this currently returning for a face at the boundary? Does neighborCentroid become the face centroid? If so, for example in 124 we need to consider the normal component instead.

Reply:
actually, I suspect these changes to a go in pair with the BC specifications. It would explain why 2nd order velocity convergence was lost when you implemented the latter

Reply:
> Does neighborCentroid become the face centroid?

If there is no neighbor, then `neighborCentroid` returns the centroid of the ghost cell...with the ghost centroid chosen such that the faceCentroid lies at the middle of the line drawn between elem and ghost centroids

Reply:
it's not orthogonal though (the line between the centroids and the surface)?
I'm trying to understand where the mods to get the jungle of equations for all BCs need to happen. My best guess is that functor? But then the distance between centroids might also need to change, since the denominator should have d_tangeant

Reply:
> it's not orthogonal though (the line between the centroids and the surface)?

You mean orthogonal to the face? No, in general the line between the element centroid and the face centroid won't be orthogonal to the face

Reply:
Ok I have dotted that distance vector with the face normal, so that term should account for non-orthogonality now.

I think any *additional* change here will be done for the purpose of skew corrections. I think the above code and all the code that underlies our INSFV simulations (framework/navier_stokes) is correct when there is no element skewness, e.g. when the face centroid is along the line segment connecting the two centroids. I have added plane-poiseuille flow tests that use interpolations to compute inlet pressure and outlet velocity, e.g. these tests do not use the `INSFVMomentum/MassAdvectionFunctionBC` objects; instead we rely on the same code that we would use in a production simulation. These plane-poiseuille tests show second order convergence for velocity whether we use average or Rhie-Chow interpolation for the velocity, and whether we use a one-term (e.g. element centroid value is boundary face value) or two-term Taylor series for the boundary face value. The pressure drops to first-order convergence when doing average velocity interpolation and one-term Taylor series for boundary face values, presumably because the implicit zero gradient assumption is incorrect for the pressure at the inlet.

This same set of physics/boundary conditions shows zeroth order convergence if I introduce significant element skewness. So for general simulations, we will definitely want to resolve #16239. I would like to resolve that in a separate PR, however, if you'll let me. This PR has gotten pretty big and the skewness corrections, e.g. trying to implement something like Moukalled 9.8 and 9.9, is going to take a fair amount of work, since now a request for `adGradSln(elem)` is actually a request for ***all*** face values and element gradients: element gradient depends on face value depends on neighbor gradient depends on neighbor face value...all the way to the edges of the mesh. So I think it would actually be a pretty big design decision deciding how to handle that. Do we do it iteratively like most codes? But we've setup our MOOSE INSFV with our traditional implicit framework...if we're going to do iterations for face values and element gradients, should we consider making segregated iterative multi-app type solves our first class capability (e.g. PIMPLE, SIMPLE, PISO), more in-line with a traditional CFD code? This is fascinating and saddening at the same time...the more times I find the recommended thing to do is to iteratively compute quantities, the more I wish we were doing DG instead.

Reply:
I think you only have to account for orthogonality on the walls? Not sure but this is the only place I've seen that term come in. You also add (1 - normal_x^2) to the numerator there. 
Did you find the projection in another equation?

Reply:
Agreed that it is skewness correction related.

Reply:
@lindsayad tagging you, this was commented on outside of the review

Reply:
> Did you find the projection in another equation?

No I made it up based on a diffusive flux dotted with the normal.

> Agreed that it is skewness correction related. If the mesh was perfectly orthogonal, the d_orth = d and n_x = 0.
> Sadly @smharper and I's meshes are not orthogonal

Moukalled seems to treat two different concepts:
1. normal and line between centroids are not collinear (non-orthogonality)
2. face centroid doesn't lie along line between cell centroids (skewness). 

I guess that in the case of a boundary face, we draw the line between the cell centroid and face centroid, so 2. is not an issue, just 1. 

I will say that I'm unsure of how much the computation of Rhie-Chow coefficients matters for our fully-coupled implicit method of solving the INS equations; more explicitly I think it doesn't matter. It seems like no matter what I do in this method, convergence comes out second order (admittedly for the orthogonal grid case). I used to only add the mass flow term if it was positive - that demonstrated second order. Then because that made the overall SNES function discontinuous, I changed to adding the mass flow no matter the sign. Still second order. In the methods outlined by Moukalled, they are using these Rhie-Chow `a` coefficients directly as their coefficients in their linearized system of equations. We are only using the coefficients in a "correction" to the velocity interpolation that ideally goes to zero as our two estimates for the pressure gradient converge.

I will try an implementation of `INSFVAdvectionBase` where I add some method like `isNoSlipWall()` and create an `INSFVWallBC` class that returns `true`. then I will modify the coefficients according to the (1-normal_comp^2) equations.

Reply:
> I will try an implementation of INSFVAdvectionBase where I add some method like isNoSlipWall() and create an INSFVWallBC class that returns true. then I will modify the coefficients according to the (1-normal_comp^2) equations.

Mmmm that's not going to work because of caching. Whatever object gets to the element centroid adjacent to the boundary first will compute the `a` coefficient first

Reply:
Yeah you are right these are two different concepts.
Ok we can try it that way if you feel it does not matter for us. How do you propose we keep track of the orthogonality correction at the boundary? Another issue? If it ends up mattering for segregated efforts, future devs will like the heads up

Reply:
I just commented out all the code in the RC coeff calculation routine and just set the coefficient to 1. Velocity convergence is still second order and the pressure convergence is actually gaining a half-order of convergence in some cases

Reply:
Well that's not a good sign. I didnt read about Rhie Chow lowering the order of convergence in the textbook
So do you still want this merged?

Reply:
I've created #16442

Reply:
> Well that's not a good sign. I didnt read about Rhie Chow lowering the order of convergence in the textbook
So do you still want this merged or do you want to look further on that

I think this is mostly pure happen-stance. I suspect that if I continued to refine the mesh we would approach the same asymptotic convergence rate. Do you know what theory says the convergence rate in the error of the pressure should be? I see either 1 or 2. Usually it is 2. I am going to take a glance and see what the differences are in the simulation setup between the 1 and 2 cases.

Reply:
Ok the difference is happening whether the MMS test is channel-flow-like (second order in pressure) or lid-driven-like (first order in pressure). I'm going to think about this a bit